components/script_plugins/unrooted_must_root.rs

```diff
@@ -87,7 +90,7 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnrootedPass {
                         cx: &LateContext,
                         def: &hir::VariantData,
                         _n: ast::Name,
-                        _gen: &hir::Generics,
+                        _gen: &hir::Generics, // use enum {generics(..), def_id(..)} instead of just def_id in UnrootedCx to (see below)
```

Comment:
.

Reply:
I don't really understand this comment -- oh, are you saying you could pass *either* a `DefId` *or` `Generics`? i.e., an enum like

```rust
enum Gen {
  Generics(Generics), DefId(DefId)
}
```

?

If so, I wouldn't bother. Calls to `generics_of` are quite cheap, no need to avoid them.

Reply:
(In particular, they are memoized)

Reply:
I could have passed the `DefId` of context, but what is the proper one here in `check_struct_def` and below in `check_variant`? I thought that this `_gen` should provide good context. That's why I came up with this enum for either one or the another. Anyway, I think getting `DefId` is the preferred way.

Reply:
@Mrowqa in `check_struct_def`, it should be the `DefId` of the struct. In `check_variant` it should be the `DefId` of the enum.